lsp/server/src/validateMobileOffline.ts

```diff
@@ -0,0 +1,141 @@
+}
+
+function traverse(node: Node, tagNames: string[], tags: Node[]): void {
+    if (node.tag && tagNames.includes(node.tag)) {
+        tags.push(node);
+    }
+
+    node.children.forEach((childNode) => traverse(childNode, tagNames, tags));
+}
+
+function findTags(document: HTMLDocument, tagNames: string[]): Node[] {
+    const tags: Node[] = [];
+    const rootNode = document.roots[0];
+
+    if (rootNode) {
+        traverse(rootNode, tagNames, tags);
+    }
+
+    return tags;
+}
+
+function camelToKebabCase(str: string): string {
+    return (
+        str
+            // Replace uppercase letters with a dash followed by the lowercase letter
+            .replace(/([A-Z])/g, '-$1')
+            // Convert the entire string to lowercase
+            .toLowerCase()
+    );
+}
+
+/**
+ * Parse componenent-experiences.yaml come in a format like this
+ *
+ * ```
+ * {
+ *   ...
+ *   'lightning:accordion': ['MobileOffline', 'Lightning', 'Communities'],
+ *   'lightning:fileUpload': ['Lightning', 'Communities'],
+ *   ...
+ * }
+ * ```
+ *
+ * This function gets array of converted keys(aka, names of base components how they are used in html),
+ * from camel case to kebob case, that doesn't have a specific string in it's array of values.
+ *
+ * @param obj Parsed component-experiences.yaml in a JSON format
+ * @param searchString A string that should not be in the key's values.
+ * @returns An array of base component names, as used in html, without a certain attribute.
+ */
+function getKeysWithoutSpecificValue(
+    obj: Record<string, string[]>,
+    searchString: string
+): string[] {
+    const result: string[] = [];
+
+    for (const [key, valueArray] of Object.entries(obj)) {
+        // Check only for LWC. They are prefixed with 'lightning:'.
+        if (
+            key.startsWith('lightning:') &&
+            !valueArray.includes(searchString)
+        ) {
+            // Base component uses '-' isntead of ':'.
+            // 'lightning' is the namespace.
+            const splittedKeys = key.split(':');
+
+            // Base component names appear as camel case in yamel. In html template,
+            // base component names use kebob case. So convert the name. For example,
+            // 'fileUpload' needs to be converted to 'file-upload'.
+            const baseComponentNameKebob = camelToKebabCase(splittedKeys[1]);
+
+            // Concatenate the namespace and component name with a dash in the middle.
+            const fixedKey = `${splittedKeys[0]}-${baseComponentNameKebob}`;
+
+            result.push(fixedKey);
+        }
+    }
+
+    return result;
+}
+
+export async function validateMobileOffline(
+    textDocument: TextDocument
+): Promise<Diagnostic[]> {
+    const nonOfflinebaseComponents = getKeysWithoutSpecificValue(
+        baseComponentsAttributes.values,
+        'MobileOffline'
+    );
+    const content = textDocument.getText();
+    const htmlDocument = parseHTMLContent(content);
```

Comment:
what will happen if parseHTMLContent throw exception when parsing invalid html? our experience on rules for js or graphql is that it crashes the lsp when failed to parse json or graphql